File: sandrose/sandrose/report/age_wise_pdc_report/age_wise_pdc_report.py
# Copyright (c) 2026, Sandrose and contributors
# For license information, please see license.txt

# import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data

import frappe
from frappe.utils import nowdate, date_diff

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": "Payment Entry", "fieldname": "payment_entry", "fieldtype": "Link", "options": "Payment Entry", "width": 200},
        {"label":"Party","fieldname":"party","fieldtype":"Data","width":150},
        {"label": "Payment Type", "fieldname": "payment_type", "fieldtype": "Data", "width": 100},
        
        {"label": "Cheque No", "fieldname": "reference_no", "fieldtype": "Data", "width": 150},
        {"label": "Cheque Date", "fieldname": "reference_date", "fieldtype": "Date", "width": 150},
        # {"label": "Cheque Issue", "fieldname": "custom_cheque_issue_date", "fieldtype": "Date", "width": 120},
        
        {"label":"Remarks","fieldname":"remarks","fieldtype":"Data","width":150},
        {"label": "Amount", "fieldname": "paid_amount", "fieldtype": "Currency", "width": 150},
        {"label": "Mode of Payment", "fieldname": "mode_of_payment", "fieldtype": "Data", "width": 150},
        {"label": "Age (Days)", "fieldname": "age", "fieldtype": "Int", "width": 100},
        {"label": "Age Bucket", "fieldname": "age_bucket", "fieldtype": "Data", "width": 120}
        
        
    ]

def get_data(filters):
    frappe.flags.ignore_cache = True
    filters = filters or {}
    # Conditions based on user inputs
    conditions = []

    # Apply filters based on user input
    if filters.get("party_type"):
        conditions.append("pe.party_type = %(party_type)s")
    if filters.get("party"):
        conditions.append("pe.party = %(party)s")
    if filters.get("payment_type"):
        conditions.append("pe.payment_type = %(payment_type)s")


    if filters.get("from_date") and filters.get("to_date"):
        conditions.append("pe.reference_date BETWEEN %(from_date)s AND %(to_date)s")

    # Join conditions with 'AND'
    conditions = " AND ".join(conditions)

    if conditions:
        conditions = " AND " + conditions

    status_map = {"Draft": 0, "Submitted": 1, "Cancelled": 2}
    docstatus = status_map.get(filters.get("status"), 1)

    params = dict(filters)
    params["docstatus"] = docstatus

    # Fetch filtered payment entries
    pdc_entries = frappe.db.sql(f"""
        SELECT
            pe.name AS payment_entry,
            pe.party,
            pe.party_name,
            pe.payment_type,
            pe.reference_no,
            pe.reference_date,
            pe.paid_amount,
            pe.mode_of_payment,

            pe.remarks
        FROM
            `tabPayment Entry` pe
        WHERE
            pe.docstatus = %(docstatus)s
            
            AND pe.reference_date IS NOT NULL
            {conditions}
        ORDER BY
            pe.reference_date ASC
    """, params, as_dict=1)

    # Process the data and calculate age
    data = []
    for entry in pdc_entries:
        age = date_diff( entry.reference_date ,nowdate())

        # Define age buckets
        if age <= 30:
            age_bucket = '0-30 Days'
        elif 31 <= age <= 60:
            age_bucket = '31-60 Days'
        elif 61 <= age <= 90:
            age_bucket = '61-90 Days'
        else:
            age_bucket = '90+ Days'

        # Append data to the list
        data.append({
            "payment_entry": entry.payment_entry,
            "party": entry.party,
            "party_name": entry.party_name,
            "payment_type": entry.payment_type,
            "reference_no": entry.reference_no,
            "reference_date": entry.reference_date,
            # "custom_cheque_issue_date":entry.custom_cheque_issue_date,
            "remarks": entry.remarks,
            "paid_amount": entry.paid_amount,
            "mode_of_payment": entry.mode_of_payment,
            "age": age,
            "age_bucket": age_bucket
        })

    return data

File: sandrose/sandrose/report/age_wise_pdc_report/age_wise_pdc_report.js
// Copyright (c) 2026, Sandrose and contributors
// For license information, please see license.txt

// frappe.query_reports["Age-wise PDC Report"] = {
// 	"filters": [

// 	]
// };

frappe.query_reports["Age-wise PDC Report"] = {
    "filters": [
		{
            "fieldname": "party_type",
            "label": __("Party Type"),
            "fieldtype": "Select",
            "options": ["Customer", "Supplier"],
            "default": "Customer",
            "reqd": 1,  // Required filter
            "on_change": function() {
                let party_type = frappe.query_report.get_filter_value('party_type');
                
                // Dynamically change the party filter based on the selected party_type
                frappe.query_report.set_filter_value('party', '');  // Clear the current party selection
                if (party_type === "Customer") {
                    frappe.query_report.get_filter('party').df.options = "Customer";
                } else if (party_type === "Supplier") {
                    frappe.query_report.get_filter('party').df.options = "Supplier";
                }

                // Force a full report refresh after setting the new filter options
                frappe.query_report.refresh();  // This triggers the report data to refresh
            }
        },
        {
            "fieldname": "party",
            "label": __("Party "),
            "fieldtype": "Link",
            "options": "Customer", // Can also be "Supplier" if needed
            "default": "",
            "reqd": 0
        },
        // {
        //     "fieldname": "party",
        //     "label": __("Party "),
        //     "fieldtype": "Link",
        //     "options": "Supplier", // Can also be "Supplier" if needed
        //     "default": "",
        //     "reqd": 0
        // },
        {
            "fieldname": "payment_type",
            "label": __("Payment Type"),
            "fieldtype": "Select",
            "options": ["Receive", "Pay", "Internal Transfer"],
            "default": "Receive",
            "reqd": 0  // Optional filter, not required
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": ["Draft", "Submitted", "Cancelled"],
            "default": "Submitted",
            "reqd": 0
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.nowdate(),  
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_days(frappe.datetime.nowdate(), 30), // Default to 30 days after today
            "reqd": 1
        }
    ],

    "onload": function(report) {
        // This will trigger when the report is loaded
        frappe.query_report.refresh();
        console.log("Report Loaded: Age Wise PDC Report");

        // You can add custom JS actions on report load here
    },

    "formatter": function(value, row, column, data, default_formatter) {
        // Custom formatter for fields if needed
        if (column.fieldname === "amount" && value) {
            // Format amount as currency, add a specific style, etc.
            value = `<span style="color:green;">${value}</span>`;
        }
        return default_formatter(value, row, column, data);
    },
    "before_load": function() {
    frappe.query_report.refresh();
    },
    
    "formatter": function(value, row, column, data, default_formatter) {
    console.log("Data row:", row, "Column:", column, "Value:", value);  // Log data for debugging
    return default_formatter(value, row, column, data);
    }


};